storage/database/sharded_database.go

```diff
@@ -155,75 +160,244 @@ func (db *shardedDB) Close() {
+func (db *shardedDB) NewIteratorUnsorted(prefix []byte, start []byte) Iterator {
+	resultCh := make(chan common.Entry, shardedDBCombineChanSize)
+	it := &shardedDBIteratorUnsorted{
+		shardedDBIterator{
+			shardedDBChanIterator: db.NewChanIterator(context.TODO(), prefix, start, resultCh),
+			resultCh:              resultCh,
+		}}
+	return it
+}
+
+// shardedDBChanIterator creates iterators for each shard DB.
+// Channels subscribing each iterators can be gained.
+// Each iterators fetch values in binary-alphabetical order.
+// This is useful when you want to operate on each items in parallel.
+type shardedDBChanIterator struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	iterators []Iterator
+
+	combinedChan bool // all workers put items to one resultChan
+	shardNum     int  // num of shards left to iterate
+	resultChs    []chan common.Entry
+}
+
+// NewChanIterator creates iterators for each shard DB. This is useful when you
+// want to operate on each items in parallel.
+// If `resultCh` is given, all items are written to `resultCh`, unsorted with a
+// particular key prefix, starting at a particular initial key. If `resultCh`
+// is not given, new channels are created for each DB. Items are written to
+// corresponding channels in binary-alphabetical order. The channels can be
+// gained by calling `Channels()`.
+//
+// If you want to get ordered items in serial, checkout shardedDB.NewIterator()
+// If you want to get unordered items in serial with Iterator Interface,
+// checkout shardedDB.NewIteratorUnsorted().
+func (db *shardedDB) NewChanIterator(ctx context.Context, prefix []byte, start []byte, resultCh chan common.Entry) shardedDBChanIterator {
+	if ctx == nil {
+		ctx = context.TODO()
+	}
+
+	it := shardedDBChanIterator{
+		ctx:          ctx,
+		cancel:       nil,
+		iterators:    make([]Iterator, len(db.shards)),
+		combinedChan: resultCh != nil,
+		shardNum:     len(db.shards),
+		resultChs:    make([]chan common.Entry, len(db.shards)),
+	}
+	it.ctx, it.cancel = context.WithCancel(ctx)
+
+	for i, shard := range db.shards {
+		it.iterators[i] = shard.NewIterator(prefix, start)
+		if resultCh == nil {
+			it.resultChs[i] = make(chan common.Entry, shardedDBSubChannelSize)
+		} else {
+			it.resultChs[i] = resultCh
+		}
+		go it.runChanWorker(it.ctx, it.iterators[i], it.resultChs[i])
+	}
+
+	return it
+}
+
+// runChanWorker runs a worker. The worker gets key/value pair from
+// `it` and push the value to `resultCh`.
+// `iterator.Release()` is called after all iterating is finished.
+// `resultCh` is closed after the iterating is finished.
+func (sit *shardedDBChanIterator) runChanWorker(ctx context.Context, it Iterator, resultCh chan common.Entry) {
+iter:
+	for it.Next() {
+		select {
+		case <-ctx.Done():
+			break iter
+		default:
+		}
+		key := make([]byte, len(it.Key()))
+		val := make([]byte, len(it.Value()))
+		copy(key, it.Key())
+		copy(val, it.Value())
```

Comment:
Why do we need to copy this?

Reply:
The data is modified when `Next()` is called. If it is not copied, the values are changed before the channel emits it.
This is a comment from `Key()` in `Iterator` interface.
> 	// Key returns the key of the current key/value pair, or nil if done. The caller
	// should not modify the contents of the returned slice, and its contents may
	// change on the next call to Next.

Reply:
Thanks. :)